packages/common/src/font.ts

```diff
@@ -136,58 +130,156 @@ export const getFontKitFont = async (textSchema: TextSchema, font: Font) => {
-  const fontSize = _fontSize || DEFAULT_FONT_SIZE;
+  // For very long lines with no whitespace use the original overPos
+  return overPosTmp > 0 ? overPosTmp : overPos - 1;
+};
+
+/**
+ * Recursively splits the line at getSplitPosition.
+ * If there is some leftover, split the rest again in the same manner.
+ */
+export const getSplittedLines = (textLine: string, calcValues: FontWidthCalcValues): string[] => {
+  const splitPos = getSplitPosition(textLine, calcValues);
+  const splittedLine = textLine.substring(0, splitPos);
+  const rest = textLine.substring(splitPos).trimStart();
+
+  if (rest.length === 0) {
+    // end recursion if there is no leftover
+    return [splittedLine];
+  }
+
+  return [splittedLine, ...getSplittedLines(rest, calcValues)];
+};
+
+/**
+ * If using dynamic font size, iteratively increment or decrement the
+ * font size to fit the containing box.
+ * Calculating space usage involves splitting lines where they exceed
+ * the box width based on the proposed size.
+ */
+export const calculateDynamicFontSize = async ({ textSchema, font, input }: { textSchema: TextSchema; font: Font; input: string; }) => {
+  const {
+    fontSize: schemaFontSize,
+    dynamicFontSize: dynamicFontSizeSetting,
+    characterSpacing: schemaCharacterSpacing,
+    width: boxWidth,
+    height: boxHeight,
+    lineHeight = DEFAULT_LINE_HEIGHT,
+  } = textSchema;
+  const fontSize = schemaFontSize || DEFAULT_FONT_SIZE;
   if (!dynamicFontSizeSetting) return fontSize;
+  if (dynamicFontSizeSetting.max < dynamicFontSizeSetting.min) return fontSize;
 
-  const characterSpacingCount = characterSpacing ?? DEFAULT_CHARACTER_SPACING;
+  const characterSpacing = schemaCharacterSpacing ?? DEFAULT_CHARACTER_SPACING;
   const fontKitFont = await getFontKitFont(textSchema, font);
   const textContentRows = input.split('\n');
 
   let dynamicFontSize = fontSize;
+  if (dynamicFontSize < dynamicFontSizeSetting.min) {
+    dynamicFontSize = dynamicFontSizeSetting.min;
+  } else if (dynamicFontSize > dynamicFontSizeSetting.max) {
+    dynamicFontSize = dynamicFontSizeSetting.max;
+  }
 
   const calculateConstraints = (size: number) => {
     let totalWidthInMm = 0;
     let totalHeightInMm = 0;
 
+    const boxWidthInPt = mm2pt(boxWidth);
     const textHeight = heightOfFontAtSize(fontKitFont, size);
-    const textHeightInMm = textHeight * DEFAULT_PT_TO_MM_RATIO;
-    textContentRows.forEach((line) => {
-      const textWidth = widthOfTextAtSize(line, fontKitFont, size);
-      const textWidthInMm = calculateTextWidthInMm(line, textWidth, characterSpacingCount);
-
-      totalWidthInMm = Math.max(totalWidthInMm, textWidthInMm);
-      totalHeightInMm += textHeightInMm;
+    const textHeightInMm = pt2mm(textHeight * lineHeight);
+
+    textContentRows.forEach((paragraph) => {
+      const lines = getSplittedLines(paragraph, {
```

Comment:
in order to get the full height of the text we need to both split it on line breaks and split by the box width